### Capstone - Cleaning Data

After extracting the data, the next section is to clean the data.

In [1]:
import pandas as pd
import re

In [2]:
df = pd.read_json('all_characters.json',orient='index')

Inspecting data

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23202 entries, 的 to 合水
Data columns (total 2 columns):
Absolute Count             23202 non-null int64
Frequency/Million words    23202 non-null float64
dtypes: float64(1), int64(1)
memory usage: 543.8+ KB


In [4]:
df

Absolute Count  Frequency/Million words
的                 4650143               45845.3460
我                 2841511               28014.2040
了                 2725964               26875.0365
不                 1752436               17277.1106
是                 1648662               16254.0119
你                 1622660               15997.6605
一                 1356666               13375.2493
在                  992131                9781.3312
有                  906700                8939.0746
个                  856510                8444.2559
就                  729449                7191.5728
人                  651396                6422.0552
要                  631789                6228.7515
好                  628627                6197.5776
都                  609802                6011.9836
也                  504107                4969.9460
和                  491847                4849.0758
想                  478573                4718.2086
还                  444030                4377.6523
啊                  437670                4314.9496
吧                  434052                4279.2800
这                  399287                3936.5350
很                  366334                3611.6543
会                  364561                3594.1745
来                  343989                3391.3569
年                  330500                3258.3701
能                  326366                3217.6133
去                  321992                3174.4905
微                  317055                3125.8170
博                  312359                3079.5196
...                   ...                      ...
树人                     19                   0.1873
译本                     19                   0.1873
无伦                     19                   0.1873
秦皇岛市                   19                   0.1873
星罗棋布                   19                   0.1873
多分                     19                   0.1873
退休证                    19                   0.1873
家闲                     19                   0.1873
巴洛克式                   19                   0.1873
cigarette              19                   0.1873
smoke                  19                   0.1873
玄关处                    19                   0.1873
都俾                     19                   0.1873
御                      19                   0.1873
怕事                     19                   0.1873
何星座                    19                   0.1873
韩范儿                    19                   0.1873
修为                     19                   0.1873
头筹                     19                   0.1873
芬芬                     19                   0.1873
玩死                     19                   0.1873
泅渡                     19                   0.1873
讓她                     19                   0.1873
阿伤                     19                   0.1873
伤心处                    19                   0.1873
Designer               19                   0.1873
安尚                     19                   0.1873
火浇油                    19                   0.1873
阿米巴                    19                   0.1873
合水                     19                   0.1873

[23202 rows x 2 columns]

We can see the dataset contains English words (Designer) and Traditional Chinese characters (讓她). Therefore is a need to clean the data.

The data cleaning process is as follows:
1. Converting Traditional Chinese characters to Simplified Chinese characters
2. Check for any duplicated Chinese characters/phrases, if there is remove them.
3. Remove any rows containing anything other than Chinese characters

Step 1: Converting Traditional Chinese to Simplified Chinese

In [5]:
from hanziconv import HanziConv

In [6]:
df.reset_index(inplace=True)
df

index  Absolute Count  Frequency/Million words
0              的         4650143               45845.3460
1              我         2841511               28014.2040
2              了         2725964               26875.0365
3              不         1752436               17277.1106
4              是         1648662               16254.0119
5              你         1622660               15997.6605
6              一         1356666               13375.2493
7              在          992131                9781.3312
8              有          906700                8939.0746
9              个          856510                8444.2559
10             就          729449                7191.5728
11             人          651396                6422.0552
12             要          631789                6228.7515
13             好          628627                6197.5776
14             都          609802                6011.9836
15             也          504107                4969.9460
16             和          491847                4849.0758
17             想          478573                4718.2086
18             还          444030                4377.6523
19             啊          437670                4314.9496
20             吧          434052                4279.2800
21             这          399287                3936.5350
22             很          366334                3611.6543
23             会          364561                3594.1745
24             来          343989                3391.3569
25             年          330500                3258.3701
26             能          326366                3217.6133
27             去          321992                3174.4905
28             微          317055                3125.8170
29             博          312359                3079.5196
...          ...             ...                      ...
23172         树人              19                   0.1873
23173         译本              19                   0.1873
23174         无伦              19                   0.1873
23175       秦皇岛市              19                   0.1873
23176       星罗棋布              19                   0.1873
23177         多分              19                   0.1873
23178        退休证              19                   0.1873
23179         家闲              19                   0.1873
23180       巴洛克式              19                   0.1873
23181  cigarette              19                   0.1873
23182      smoke              19                   0.1873
23183        玄关处              19                   0.1873
23184         都俾              19                   0.1873
23185          御              19                   0.1873
23186         怕事              19                   0.1873
23187        何星座              19                   0.1873
23188        韩范儿              19                   0.1873
23189         修为              19                   0.1873
23190         头筹              19                   0.1873
23191         芬芬              19                   0.1873
23192         玩死              19                   0.1873
23193         泅渡              19                   0.1873
23194         讓她              19                   0.1873
23195         阿伤              19                   0.1873
23196        伤心处              19                   0.1873
23197   Designer              19                   0.1873
23198         安尚              19                   0.1873
23199        火浇油              19                   0.1873
23200        阿米巴              19                   0.1873
23201         合水              19                   0.1873

[23202 rows x 3 columns]

In [7]:
idx = pd.Series(df['index'])
idx_simple = idx.map(lambda x: HanziConv.toSimplified(x))
idx_simple

0                的
1                我
2                了
3                不
4                是
5                你
6                一
7                在
8                有
9                个
10               就
11               人
12               要
13               好
14               都
15               也
16               和
17               想
18               还
19               啊
20               吧
21               这
22               很
23               会
24               来
25               年
26               能
27               去
28               微
29               博
           ...    
23172           树人
23173           译本
23174           无伦
23175         秦皇岛市
23176         星罗棋布
23177           多分
23178          退休证
23179           家闲
23180         巴洛克式
23181    cigarette
23182        smoke
23183          玄关处
23184           都俾
23185            御
23186           怕事
23187          何星座
23188          韩范儿
23189           修为
23190           头筹
23191           芬芬
23192           玩死
23193       

In [8]:
df['index_simple'] = idx_simple
df.drop(columns='index',inplace=True)
df

Absolute Count  Frequency/Million words index_simple
0             4650143               45845.3460            的
1             2841511               28014.2040            我
2             2725964               26875.0365            了
3             1752436               17277.1106            不
4             1648662               16254.0119            是
5             1622660               15997.6605            你
6             1356666               13375.2493            一
7              992131                9781.3312            在
8              906700                8939.0746            有
9              856510                8444.2559            个
10             729449                7191.5728            就
11             651396                6422.0552            人
12             631789                6228.7515            要
13             628627                6197.5776            好
14             609802                6011.9836            都
15             504107                4969.9460            也
16             491847                4849.0758            和
17             478573                4718.2086            想
18             444030                4377.6523            还
19             437670                4314.9496            啊
20             434052                4279.2800            吧
21             399287                3936.5350            这
22             366334                3611.6543            很
23             364561                3594.1745            会
24             343989                3391.3569            来
25             330500                3258.3701            年
26             326366                3217.6133            能
27             321992                3174.4905            去
28             317055                3125.8170            微
29             312359                3079.5196            博
...               ...                      ...          ...
23172              19                   0.1873           树人
23173              19                   0.1873           译本
23174              19                   0.1873           无伦
23175              19                   0.1873         秦皇岛市
23176              19                   0.1873         星罗棋布
23177              19                   0.1873           多分
23178              19                   0.1873          退休证
23179              19                   0.1873           家闲
23180              19                   0.1873         巴洛克式
23181              19                   0.1873    cigarette
23182              19                   0.1873        smoke
23183              19                   0.1873          玄关处
23184              19                   0.1873           都俾
23185              19                   0.1873            御
23186              19                   0.1873           怕事
23187              19                   0.1873          何星座
23188              19                   0.1873          韩范儿
23189              19                   0.1873           修为
23190              19                   0.1873           头筹
23191              19                   0.1873           芬芬
23192              19                   0.1873           玩死
23193              19                   0.1873           泅渡
23194              19                   0.1873           让她
23195              19                   0.1873           阿伤
23196              19                   0.1873          伤心处
23197              19                   0.1873     Designer
23198              19                   0.1873           安尚
23199              19                   0.1873          火浇油
23200              19                   0.1873          阿米巴
23201              19                   0.1873           合水

[23202 rows x 3 columns]

Step 2: Checking for any duplicated rows. We create a new dataframe by dropping the rows with duplicated indexes (if there are any).

In [9]:
df_new = df.drop_duplicates(subset='index_simple',keep='first')
df_new

Absolute Count  Frequency/Million words index_simple
0             4650143               45845.3460            的
1             2841511               28014.2040            我
2             2725964               26875.0365            了
3             1752436               17277.1106            不
4             1648662               16254.0119            是
5             1622660               15997.6605            你
6             1356666               13375.2493            一
7              992131                9781.3312            在
8              906700                8939.0746            有
9              856510                8444.2559            个
10             729449                7191.5728            就
11             651396                6422.0552            人
12             631789                6228.7515            要
13             628627                6197.5776            好
14             609802                6011.9836            都
15             504107                4969.9460            也
16             491847                4849.0758            和
17             478573                4718.2086            想
18             444030                4377.6523            还
19             437670                4314.9496            啊
20             434052                4279.2800            吧
21             399287                3936.5350            这
22             366334                3611.6543            很
23             364561                3594.1745            会
24             343989                3391.3569            来
25             330500                3258.3701            年
26             326366                3217.6133            能
27             321992                3174.4905            去
28             317055                3125.8170            微
29             312359                3079.5196            博
...               ...                      ...          ...
23172              19                   0.1873           树人
23173              19                   0.1873           译本
23174              19                   0.1873           无伦
23175              19                   0.1873         秦皇岛市
23176              19                   0.1873         星罗棋布
23177              19                   0.1873           多分
23178              19                   0.1873          退休证
23179              19                   0.1873           家闲
23180              19                   0.1873         巴洛克式
23181              19                   0.1873    cigarette
23182              19                   0.1873        smoke
23183              19                   0.1873          玄关处
23184              19                   0.1873           都俾
23185              19                   0.1873            御
23186              19                   0.1873           怕事
23187              19                   0.1873          何星座
23188              19                   0.1873          韩范儿
23189              19                   0.1873           修为
23190              19                   0.1873           头筹
23191              19                   0.1873           芬芬
23192              19                   0.1873           玩死
23193              19                   0.1873           泅渡
23194              19                   0.1873           让她
23195              19                   0.1873           阿伤
23196              19                   0.1873          伤心处
23197              19                   0.1873     Designer
23198              19                   0.1873           安尚
23199              19                   0.1873          火浇油
23200              19                   0.1873          阿米巴
23201              19                   0.1873           合水

[22769 rows x 3 columns]

The number of rows has decreased to 22769. There are about 500 rows removed.

Step 3: Remove any rows containing non-chinese characters. The unicode range for Chinese characters is referred from this [PDF](http://www.unicode.org/versions/Unicode5.0.0/ch12.pdf#G12159)

In [36]:
df_chinese = pd.Series(df_new['index_simple'])
df_chinese_only = pd.Series(df_chinese[df_chinese.str.match('[\u4e00-\u9FFF]') == True])
df_cleaned = df.iloc[df_chinese_only.index]
df_cleaned

Absolute Count  Frequency/Million words index_simple
0             4650143               45845.3460            的
1             2841511               28014.2040            我
2             2725964               26875.0365            了
3             1752436               17277.1106            不
4             1648662               16254.0119            是
5             1622660               15997.6605            你
6             1356666               13375.2493            一
7              992131                9781.3312            在
8              906700                8939.0746            有
9              856510                8444.2559            个
10             729449                7191.5728            就
11             651396                6422.0552            人
12             631789                6228.7515            要
13             628627                6197.5776            好
14             609802                6011.9836            都
15             504107                4969.9460            也
16             491847                4849.0758            和
17             478573                4718.2086            想
18             444030                4377.6523            还
19             437670                4314.9496            啊
20             434052                4279.2800            吧
21             399287                3936.5350            这
22             366334                3611.6543            很
23             364561                3594.1745            会
24             343989                3391.3569            来
25             330500                3258.3701            年
26             326366                3217.6133            能
27             321992                3174.4905            去
28             317055                3125.8170            微
29             312359                3079.5196            博
...               ...                      ...          ...
23169              19                   0.1873          汽油味
23170              19                   0.1873         东施效颦
23171              19                   0.1873           烧包
23172              19                   0.1873           树人
23173              19                   0.1873           译本
23174              19                   0.1873           无伦
23175              19                   0.1873         秦皇岛市
23176              19                   0.1873         星罗棋布
23177              19                   0.1873           多分
23178              19                   0.1873          退休证
23179              19                   0.1873           家闲
23180              19                   0.1873         巴洛克式
23183              19                   0.1873          玄关处
23184              19                   0.1873           都俾
23185              19                   0.1873            御
23186              19                   0.1873           怕事
23187              19                   0.1873          何星座
23188              19                   0.1873          韩范儿
23189              19                   0.1873           修为
23190              19                   0.1873           头筹
23191              19                   0.1873           芬芬
23192              19                   0.1873           玩死
23193              19                   0.1873           泅渡
23194              19                   0.1873           让她
23195              19                   0.1873           阿伤
23196              19                   0.1873          伤心处
23198              19                   0.1873           安尚
23199              19                   0.1873          火浇油
23200              19                   0.1873          阿米巴
23201              19                   0.1873           合水

[21206 rows x 3 columns]

The dataset is now cleaned. Now we have to change the index to the chinese characters/phrases and save the file as a CSV.

In [38]:
df_cleaned.rename(columns={'index_simple':'Index'},inplace=True)
df_cleaned.set_index('Index',inplace=True)

C:\Users\Lenovo\Anaconda3\lib\site-packages\pandas\core\frame.py:3781: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return super(DataFrame, self).rename(**kwargs)


In [39]:
df_cleaned

Absolute Count  Frequency/Million words
Index                                         
的             4650143               45845.3460
我             2841511               28014.2040
了             2725964               26875.0365
不             1752436               17277.1106
是             1648662               16254.0119
你             1622660               15997.6605
一             1356666               13375.2493
在              992131                9781.3312
有              906700                8939.0746
个              856510                8444.2559
就              729449                7191.5728
人              651396                6422.0552
要              631789                6228.7515
好              628627                6197.5776
都              609802                6011.9836
也              504107                4969.9460
和              491847                4849.0758
想              478573                4718.2086
还              444030                4377.6523
啊              437670                4314.9496
吧              434052                4279.2800
这              399287                3936.5350
很              366334                3611.6543
会              364561                3594.1745
来              343989                3391.3569
年              330500                3258.3701
能              326366                3217.6133
去              321992                3174.4905
微              317055                3125.8170
博              312359                3079.5196
...               ...                      ...
汽油味                19                   0.1873
东施效颦               19                   0.1873
烧包                 19                   0.1873
树人                 19                   0.1873
译本                 19                   0.1873
无伦                 19                   0.1873
秦皇岛市               19                   0.1873
星罗棋布               19                   0.1873
多分                 19                   0.1873
退休证                19                   0.1873
家闲                 19                   0.1873
巴洛克式               19                   0.1873
玄关处                19                   0.1873
都俾                 19                   0.1873
御                  19                   0.1873
怕事                 19                   0.1873
何星座                19                   0.1873
韩范儿                19                   0.1873
修为                 19                   0.1873
头筹                 19                   0.1873
芬芬                 19                   0.1873
玩死                 19                   0.1873
泅渡                 19                   0.1873
让她                 19                   0.1873
阿伤                 19                   0.1873
伤心处                19                   0.1873
安尚                 19                   0.1873
火浇油                19                   0.1873
阿米巴                19                   0.1873
合水                 19                   0.1873

[21206 rows x 2 columns]

In [43]:
df_cleaned.to_csv('cleaned.csv',index=True)